[MTIA] [PyTorch] [Export] Allow setting names for lifted constants before tracing.

patrick-toulme opened the pull request:
Summary:
The lift_constants_pass will set the name of lifted to constants to module_path_{lifted_tensor_counter}. This makes it very difficult for a compiler or runtime to map certain constants to what the placeholder value is in the IR. 

This PR adds a new API to Torch Export allowing users to set the names of certain tensors. The lift_constants_pass then respects this name. 

This is not supported for strict=True or for naming FakeTensors.

```
        tensor = torch.tensor([1, 2, 3])
        result = torch.export._name_tensor_constants.set_tensor_name(tensor, "my_custom_tensor")
        self.assertIs(result, tensor)  # Should return the same tensor
```

Before:
```
 %model_layers_0_self_attn_attn_lifted_tensor_0 : bf16[2, 346732, 16, 8, 64][num_users=2] = placeholder[target='model_layers_0_self_attn_attn_lifted_tensor_0']
```

```
 kv_cache = torch.export._name_tensor_constants.set_tensor_name(kv_cache, "kv_cache")
```

After:


```
%model_layers_0_self_attn_attn_kv_cache : bf16[2, 346544, 16, 8, 64][num_users=2] = placeholder[target='model_layers_0_self_attn_attn_kv_cache']
```

Test Plan:
Added unit tests

Rollback Plan:

Differential Revision: D82387095

jansel reviewed (changes requested):
Failing tests?

angelayi reviewed (commented):
I believe we only name things "...lifted_tensor" when there is a `torch.tensor(..)` call in the graph, not when it is set as an attribute on init, so it's not clear to me what issue you're running into. Could you check your test cases?

This example should have a constant called `c_constant_tensor`. 
```
        class SimpleModule(torch.nn.Module):
            def __init__(self):
                super().__init__()
                # Create a named tensor that will be lifted
                self.constant_tensor = torch.tensor([10, 20, 30])

            def forward(self, x):
                return x + self.constant_tensor
```

Whereas this example should have a constant called `c_lifted_tensor_0`:
```
        class SimpleModule(torch.nn.Module):
            def __init__(self):
                super().__init__()

            def forward(self, x):
                return x + torch.tensor(1)
```

patrick-toulme commented:
@angelayi  I will send you my workload internally to show you why I need this

angelayi reviewed (dismissed):
I'm not opposed to theoretically adding an api to hint to export the names for these tensor constants, but I don't think this works for export(..., strict=True). An easier fix would be to rewrite the original code to register the constant as a buffer.

patrick-toulme commented on torch/export/__init__.py:
```diff
@@ -53,6 +56,76 @@
 
+
+# Tensor naming utilities for torch export
+def set_tensor_name(tensor: torch.Tensor, name: str) -> torch.Tensor:
```
will fix this

patrick-toulme commented on docs/source/torch.rst:
```diff
@@ -768,6 +768,16 @@ Export Path
+
+    export.get_tensor_name
+    export.name
+    export.set_tensor_name
```
will fix

patrick-toulme commented on test/export/test_tensor_naming.py:
```diff
@@ -0,0 +1,180 @@
+        exported_output = exported_program.module()(input_tensor)
+
+        # Results should be identical
+        torch.testing.assert_close(original_output, exported_output)
```
Will make better test cases and use torch.export.export

patrick-toulme commented:
> I'm not opposed to theoretically adding an api to hint to export the names for these tensor constants, but I don't think this works for export(..., strict=True). An easier fix would be to rewrite the original code to register the constant as a buffer.

as we discussed, I am not allowed to rewrite vLLM code and it would be a major rewrite of vLLM. I will make the other changes you requested. Thank you!

zhxchen17 reviewed (dismissed):
Not sure about adding a private interface that will only benefit 1 group of torch.export user, especially when node renaming is doable out of export.

This can get complicated for maintenance in the long term, when tensor subclasses are involved (which decomposes a tensor and break name convention a user would like to establish here). This is also fragile if we re-export the original module in processing pipeline which can easily screw up the node names.

A better approach is to add separate metadata to placeholder nodes indicating the "sources" of input tensors (can be dynamo sources or whatever) regardless of annotating tensors. This is more likely to scale to different use cases and useful for debugging purposes anyway.

patrick-toulme commented:
@zhxchen17  Let's discuss internally. This is blocking any further development of vLLM for MTIA graph mode.

patrick-toulme commented:
> Not sure about adding a private interface that will only benefit 1 group of torch.export user, especially when node renaming is doable out of export.
> 
> This can get complicated for maintenance in the long term, when tensor subclasses are involved (which decomposes a tensor and break name convention a user would like to establish here). This is also fragile if we re-export the original module in processing pipeline which can easily screw up the node names.
> 
> A better approach is to add separate metadata to placeholder nodes indicating the "sources" of input tensors (can be dynamo sources or whatever) regardless of annotating tensors. This is more likely to scale to different use cases and useful for debugging purposes anyway.

Let me add more context to this PR, 

1. MTIA is not using dynamo, we are using export(strict=False) in our compiler. 

We are in a situation where vLLM creates raw tensors for a KV cache here 

https://github.com/vllm-project/vllm/blob/e3db5ebb66590031ecfd3338de41b6f1ee95bf2a/vllm/v1/worker/gpu_model_runner.py#L3581

This is a torch.zeros() call outside of our jitted graph at model init time. 


We then jit the entire forward of the model to make one graph for the entire model. 

The KV Cache is of shape [2, 346732, 16, 8, 64]. 


vLLM calls .unbind() in the forward graph

https://github.com/vllm-project/vllm/blob/e3db5ebb66590031ecfd3338de41b6f1ee95bf2a/vllm/v1/attention/backends/xformers.py#L356

This decomposes into 
```
kv_cache = lifted_constant(kv_cache)
key = slice(kv_cache)
reshaped_key = reshape(slice)
```

Moving on to MTIA runtime / compiler:

1. We support in-place writes to graph inputs (kv_cache is a graph input here). Our compiler internally will transform our lower end internal dialect to write in-place.
2. The user has to tell us the targets of placeholders that the user would like us to write in-place too
3. In our runtime, the user has to provide us a raw map of placeholder target to raw tensor object. 

In the current form of lifted_tensor_# name, the user has no way to pass to MTIA Runtime/Compiler the map of placeholder targets to raw tensors. 


This PR is quite small and maintains existing behavior of PyTorch. If we cannot land, this I really need PyTorch to provide some solution to my problem above. 

Also, we are not allowed to modify vLLM code to use register_buffer() etc. That would be a massive refactor of vLLM.

jansel commented on torch/_export/passes/lift_constants_pass.py:
```diff
@@ -274,17 +275,35 @@ def lift_constants_pass(
                         constant_name = f"lifted_tensor_{num_tensor_constants}"
                         constant_fqn = get_constant_fqn(node, constant_name)
-                    num_tensor_constants += 1
+                        while constant_fqn in used_target_names:
+                            num_tensor_constants += 1
+                            constant_name = f"lifted_tensor_{num_tensor_constants}"
+                            constant_fqn = get_constant_fqn(node, constant_name)
+                        num_tensor_constants += 1
```
Duplicate code with the above?

I'd suggest having a single helper to get the name, rather than both `get_tensor_name` and `get_constant_fqn`.   And remove the `if custom_name is not None` branch from this function.

patrick-toulme commented:
@pytorchbot merge

zhxchen17 reviewed (changes requested):
sorry, my concerns not addressed

patrick-toulme committed:
[MTIA] [PyTorch] [Export] Allow setting names for lifted constants before tracing. (#162926)

Summary:

The lift_constants_pass will set the name of lifted to constants to module_path_{lifted_tensor_counter}. This makes it very difficult for a compiler or runtime to map certain constants to what the placeholder value is in the IR. 

This PR adds a new API to Torch Export allowing users to set the names of certain tensors. The lift_constants_pass then respects this name. 

This is not supported for strict=True or for naming FakeTensors.

```
        tensor = torch.tensor([1, 2, 3])
        result = torch.export._name_tensor_constants.set_tensor_name(tensor, "my_custom_tensor")
        self.assertIs(result, tensor)  # Should return the same tensor
```

Before:
```
 %model_layers_0_self_attn_attn_lifted_tensor_0 : bf16[2, 346732, 16, 8, 64][num_users=2] = placeholder[target='model_layers_0_self_attn_attn_lifted_tensor_0']
```

```
 kv_cache = torch.export._name_tensor_constants.set_tensor_name(kv_cache, "kv_cache")
```

After:


```
%model_layers_0_self_attn_attn_kv_cache : bf16[2, 346544, 16, 8, 64][num_users=2] = placeholder[target='model_layers_0_self_attn_attn_kv_cache']
```

Test Plan: Added unit tests

Differential Revision: D82387095

angelayi reviewed (changes requested):
+1 my concerns were also not addressed :) also I believe we've found a better solution in https://github.com/pytorch/pytorch/pull/163236
